pkg/ds/farm.go

```diff
@@ -0,0 +1,342 @@
+				if podLabels.Labels.Has(DSIDLabel) {
+					dsID := podLabels.Labels.Get(DSIDLabel)
+
+					// Check if the daemon set exists, if it doesn't unschedule the pod
+					if _, ok := dsIDMap[dsID]; !ok {
+						nodeName, podID, err := labels.NodeAndPodIDFromPodLabel(podLabels)
+						if err != nil {
+							dsf.logger.NoFields().Fatal(err)
+							return
+						}
+
+						// TODO: Since this mirrors the unschedule function in daemon_set.go,
+						// We should find a nice way to couple them together
+						dsf.logger.NoFields().Infof("Unscheduling from %s with dangling daemon set uuid %s", nodeName, dsID)
+
+						_, err = dsf.kpStore.DeletePod(kp.INTENT_TREE, nodeName, podID)
+						if err != nil {
+							dsf.logger.NoFields().Errorf("Unable to delete pod from intent tree: %v", err)
+						}
+
+						id := labels.MakePodLabelKey(nodeName, podID)
+						err = dsf.applicator.RemoveLabel(labels.POD, id, DSIDLabel)
+						if err != nil {
+							dsf.logger.NoFields().Errorf("Error removing label: %v", err)
+						}
+					}
+				}
+			}
+		}
+	}()
+}
+
+func (dsf *Farm) mainLoop(quitCh <-chan struct{}) {
+	subQuit := make(chan struct{})
+	defer close(subQuit)
+	dsWatch := dsf.dsStore.Watch(subQuit)
+
+	var changes *dsstore.WatchedDaemonSets
+	for {
+		select {
+		case <-quitCh:
+			return
+		case val, ok := <-dsWatch:
+			if !ok {
+				return
+			}
+			changes = val
+			if changes == nil {
+				dsf.logger.Error("Unexpected nil value received on channel")
+				return
+			}
+		}
+
+		for _, child := range dsf.children {
+			select {
+			case <-quitCh:
+				return
+			case err, ok := <-child.errCh:
+				if !ok {
+					// child error channel closed
+					return
+				}
+				dsf.logger.Fatalf("An fatal error has occured in ds: '%v':, %v", child.ds, err)
+			default:
+			}
+		}
+
+		err := dsf.handleDSChanges(quitCh, changes)
+		if err != nil {
+			dsf.logger.Fatalf("%v", err)
+			return
+		}
+	}
+}
+
+func (dsf *Farm) handleDSChanges(quitCh <-chan struct{}, changes *dsstore.WatchedDaemonSets) error {
+	if changes.Err != nil {
+		dsf.logger.Infof("An error has occrred while watching daemon sets: %v", changes.Err)
+	}
+
+	if len(changes.Created) > 0 {
+		dsf.logger.Infof("The following daemon sets have been created:")
+		for _, dsFields := range changes.Created {
+			dsf.logger.Infof("%v", *dsFields)
+		}
+
+		for _, dsFields := range changes.Created {
+			// If the daemon set contends with another daemon set, disable it
+			// if that fails, delete it, if that fails too this is a fatal error
+			dsIDContended, isContended, err := dsf.dsContends(dsFields)
+			if err != nil {
+				return util.Errorf("Fatal error occured when trying to check for daemon set contention: %v", err)
+			}
+
+			if isContended {
+				dsf.logger.Errorf("Created daemon set '%s' contends with %s", dsFields.ID, dsIDContended)
+				newDS, err := dsf.dsStore.Disable(dsFields.ID)
+				if err != nil {
+					return util.Errorf("Fatal error occured when trying to delete daemon set: %v", err)
```

Comment:
I think that this means any contended DaemonSet can stop the progress of all other DaemonSets. Would a `continue` statement be more robust?